***************
*** 21,24 ****
--- 21,32 ----
   #pragma alloca
  #endif
+ #ifdef __MVS__
+   #ifndef alloca                                           
+     #define alloca(x) __alloca(x)                          
+     #pragma linkage(__alloca,builtin)                    
+     void *__alloca(unsigned int x);                      
+     #define __alloca __alloca
+   #endif
+ #endif
  
  #ifdef	HAVE_CONFIG_H
***************
*** 230,234 ****
  
  #ifndef __GNU_LIBRARY__
! # define __stat stat
  # ifdef STAT_MACROS_BROKEN
  #  undef S_ISDIR
--- 238,242 ----
  
  #ifndef __GNU_LIBRARY__
! # define __gnustat stat
  # ifdef STAT_MACROS_BROKEN
  #  undef S_ISDIR
***************
*** 248,253 ****
  # define getpwnam_r(name, bufp, buf, len, res) \
     __getpwnam_r (name, bufp, buf, len, res)
! # ifndef __stat
! #  define __stat(fname, buf) __xstat (_STAT_VER, fname, buf)
  # endif
  #endif
--- 256,261 ----
  # define getpwnam_r(name, bufp, buf, len, res) \
     __getpwnam_r (name, bufp, buf, len, res)
! # ifndef __gnustat
! #  define __gnustat(fname, buf) __xstat (_STAT_VER, fname, buf)
  # endif
  #endif
***************
*** 285,290 ****
  #endif
  
! #if !defined __stat
! # define __stat stat
  #endif
  
--- 293,298 ----
  #endif
  
! #if !defined __gnustat
! # define __gnustat stat
  #endif
  
***************
*** 817,821 ****
  	  || (((flags & GLOB_ALTDIRFUNC)
  	       ? (*pglob->gl_stat) (dirname, &st)
! 	       : __stat (dirname, &st)) == 0
  	      && S_ISDIR (st.st_mode)))
  	{
--- 825,829 ----
  	  || (((flags & GLOB_ALTDIRFUNC)
  	       ? (*pglob->gl_stat) (dirname, &st)
! 	       : __gnustat (dirname, &st)) == 0
  	      && S_ISDIR (st.st_mode)))
  	{
***************
*** 963,967 ****
  		  /* First check whether this really is a directory.  */
  		  if (((flags & GLOB_ALTDIRFUNC)
! 		       ? (*pglob->gl_stat) (dir, &st) : __stat (dir, &st)) != 0
  		      || !S_ISDIR (st.st_mode))
  		    /* No directory, ignore this entry.  */
--- 971,975 ----
  		  /* First check whether this really is a directory.  */
  		  if (((flags & GLOB_ALTDIRFUNC)
! 		       ? (*pglob->gl_stat) (dir, &st) : __gnustat (dir, &st)) != 0
  		      || !S_ISDIR (st.st_mode))
  		    /* No directory, ignore this entry.  */
***************
*** 1039,1043 ****
  	if (((flags & GLOB_ALTDIRFUNC)
  	     ? (*pglob->gl_stat) (pglob->gl_pathv[i], &st)
! 	     : __stat (pglob->gl_pathv[i], &st)) == 0
  	    && S_ISDIR (st.st_mode))
  	  {
--- 1047,1051 ----
  	if (((flags & GLOB_ALTDIRFUNC)
  	     ? (*pglob->gl_stat) (pglob->gl_pathv[i], &st)
! 	     : __gnustat (pglob->gl_pathv[i], &st)) == 0
  	    && S_ISDIR (st.st_mode))
  	  {
***************
*** 1272,1276 ****
  	  if (((flags & GLOB_ALTDIRFUNC)
  	       ? (*pglob->gl_stat) (fullname, &st)
! 	       : __stat (fullname, &st)) == 0)
  	    /* We found this file to be existing.  Now tell the rest
  	       of the function to copy this name into the result.  */
--- 1280,1284 ----
  	  if (((flags & GLOB_ALTDIRFUNC)
  	       ? (*pglob->gl_stat) (fullname, &st)
! 	       : __gnustat (fullname, &st)) == 0)
  	    /* We found this file to be existing.  Now tell the rest
  	       of the function to copy this name into the result.  */
