***************
*** 6695,6699 ****
  
  
! ALLOCA=\${LIBOBJDIR}alloca.$ac_objext
  
  $as_echo "#define C_ALLOCA 1" >>confdefs.h
--- 6695,6699 ----
  
  
! ALLOCA=lib/alloca.$ac_objext
  
  $as_echo "#define C_ALLOCA 1" >>confdefs.h
***************
*** 18053,18058 ****
    # 2. Prepend LIBOBJDIR.  When used with automake>=1.10 LIBOBJDIR
    #    will be set to the directory where LIBOBJS objects are built.
!   as_fn_append ac_libobjs " \${LIBOBJDIR}$ac_i\$U.$ac_objext"
!   as_fn_append ac_ltlibobjs " \${LIBOBJDIR}$ac_i"'$U.lo'
  done
  LIBOBJS=$ac_libobjs
--- 18053,18058 ----
    # 2. Prepend LIBOBJDIR.  When used with automake>=1.10 LIBOBJDIR
    #    will be set to the directory where LIBOBJS objects are built.
!   as_fn_append ac_libobjs " /lib/$ac_i\$U.$ac_objext"
!   as_fn_append ac_ltlibobjs " /lib/$ac_i"'$U.lo'
  done
  LIBOBJS=$ac_libobjs
