***************
*** 28,31 ****
--- 28,38 ----
  #include "filename.h"
  
+ #ifdef __MVS__
+   /* msf - inefficient hack for now */
+   char* stpcpy(char* dest, const char* src) {
+     strcpy(dest,src);
+     return &dest[strlen(dest)];
+   }
+ #endif
  /* Concatenate a directory filename, a relative filename and an optional
     suffix.  The directory may end with the directory separator.  The second
