***************
*** 152,157 ****
  /* Match STRING against the filename pattern PATTERN, returning zero if
     it matches, nonzero if not.  */
! static int internal_fnmatch __P ((const char *pattern, const char *string,
! 				  int no_leading_period, int flags))
       internal_function;
  static int
--- 152,157 ----
  /* Match STRING against the filename pattern PATTERN, returning zero if
     it matches, nonzero if not.  */
! static int internal_fnmatch (const char *pattern, const char *string,
! 				  int no_leading_period, int flags)
       internal_function;
  static int
