***************
*** 70,74 ****
  # use ar to extract alloca.o from them instead of compiling alloca.c.
  AC_LIBSOURCES(alloca.c)
! AC_SUBST([ALLOCA], [\${LIBOBJDIR}alloca.$ac_objext])dnl
  AC_DEFINE(C_ALLOCA, 1, [Define to 1 if using 'alloca.c'.])
  
--- 70,74 ----
  # use ar to extract alloca.o from them instead of compiling alloca.c.
  AC_LIBSOURCES(alloca.c)
! AC_SUBST([ALLOCA], [alloca.$ac_objext])dnl
  AC_DEFINE(C_ALLOCA, 1, [Define to 1 if using 'alloca.c'.])
  
